[FX] Support tensor-valued constants

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43741 [FX][RFC] Input prototype for collection types in sym trace
* #43740 [FX] Better error when unpacking Proxy
* #43674 [FX] Pickle serialization of GraphModule via forward source
* **#43666 [FX] Support tensor-valued constants**

Tensors do not have a reliable string repr() from which they can be
constructed (and we probably don't want to rely on that, either), so
for any constant Tensor values we encounter, first search for if they
are an attribute of some module in the module hierarchy. If so, emit
a get_param to retrieve that tensor. Otherwise, we'll store away the
tensor value into a special attribute on the Module s.t. we can
retrieve it with a get_param.

Closes https://github.com/pytorch/pytorch/issues/43512

Differential Revision: [D23359110](https://our.internmc.facebook.com/intern/diff/D23359110)

Commit:
[FX] Support tensor-valued constants

[ghstack-poisoned]

Commit:
Update on "[FX] Support tensor-valued constants"


Differential Revision: [D23359110](https://our.internmc.facebook.com/intern/diff/D23359110)

[ghstack-poisoned]

jamesr66a commented on torch/fx/symbolic_trace.py:
```diff
@@ -106,6 +106,44 @@ def create_arg(self, a: Any) -> Argument:
+                i = 0
+                while True:
+                    qualname = f'__tensor_constant{i}'
+                    if not hasattr(self.root, qualname):
```
I don't like how this is being dumped on `root`. Ideally it would be somewhere else, like on the produced GraphModule. @zdevito @suo any suggestions?

dzhulgakov replied:
it should be naturally addressed if you eliminate `root` and instead copy submodules (while filtering only the ones that occur in `get_param`). We talked about it all together but I'm not sure whether anyone is working on it (@zdevito ?)

Commit:
Update on "[FX] Support tensor-valued constants"


Tensors do not have a reliable string repr() from which they can be
constructed (and we probably don't want to rely on that, either), so
for any constant Tensor values we encounter, first search for if they
are an attribute of some module in the module hierarchy. If so, emit
a get_param to retrieve that tensor. Otherwise, we'll store away the
tensor value into a special attribute on the Module s.t. we can
retrieve it with a get_param.

Differential Revision: [D23359110](https://our.internmc.facebook.com/intern/diff/D23359110)

[ghstack-poisoned]

Commit:
Update on "[FX] Support tensor-valued constants"


Tensors do not have a reliable string repr() from which they can be
constructed (and we probably don't want to rely on that, either), so
for any constant Tensor values we encounter, first search for if they
are an attribute of some module in the module hierarchy. If so, emit
a get_param to retrieve that tensor. Otherwise, we'll store away the
tensor value into a special attribute on the Module s.t. we can
retrieve it with a get_param.

Closes https://github.com/pytorch/pytorch/issues/43512

Differential Revision: [D23359110](https://our.internmc.facebook.com/intern/diff/D23359110)

[ghstack-poisoned]

dzhulgakov commented on torch/fx/symbolic_trace.py:
```diff
@@ -106,6 +106,44 @@ def create_arg(self, a: Any) -> Argument:
+                    if isinstance(p, torch.Tensor) and a is p:
+                        return '.'.join(qualname_atoms + [n])
+                for n, c in m.named_children():
+                    maybe_result : Optional[str] = search_for_tensor(c, qualname_atoms + [n])
```
it's really minor, but why pass down `qualname_atoms` instead of just prepending it on the return two lines below?

dzhulgakov commented on torch/fx/symbolic_trace.py:
```diff
@@ -106,6 +106,44 @@ def create_arg(self, a: Any) -> Argument:
+                child submodules. If it's not found there, return None
+                """
+                for n, p in m.__dict__.items():
+                    if isinstance(p, torch.Tensor) and a is p:
```
I think just `a is p` is sufficient (`is` is not overridable afaiu)

Commit:
Update on "[FX] Support tensor-valued constants"


Tensors do not have a reliable string repr() from which they can be
constructed (and we probably don't want to rely on that, either), so
for any constant Tensor values we encounter, first search for if they
are an attribute of some module in the module hierarchy. If so, emit
a get_param to retrieve that tensor. Otherwise, we'll store away the
tensor value into a special attribute on the Module s.t. we can
retrieve it with a get_param.

Closes https://github.com/pytorch/pytorch/issues/43512

Differential Revision: [D23359110](https://our.internmc.facebook.com/intern/diff/D23359110)

[ghstack-poisoned]

Commit:
Update on "[FX] Support tensor-valued constants"


Tensors do not have a reliable string repr() from which they can be
constructed (and we probably don't want to rely on that, either), so
for any constant Tensor values we encounter, first search for if they
are an attribute of some module in the module hierarchy. If so, emit
a get_param to retrieve that tensor. Otherwise, we'll store away the
tensor value into a special attribute on the Module s.t. we can
retrieve it with a get_param.

Closes https://github.com/pytorch/pytorch/issues/43512

Differential Revision: [D23359110](https://our.internmc.facebook.com/intern/diff/D23359110)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
This can be simpler if we get rid of `root`, but it can be done later

Commit:
Update on "[FX] Support tensor-valued constants"


Tensors do not have a reliable string repr() from which they can be
constructed (and we probably don't want to rely on that, either), so
for any constant Tensor values we encounter, first search for if they
are an attribute of some module in the module hierarchy. If so, emit
a get_param to retrieve that tensor. Otherwise, we'll store away the
tensor value into a special attribute on the Module s.t. we can
retrieve it with a get_param.

Closes https://github.com/pytorch/pytorch/issues/43512

Differential Revision: [D23359110](https://our.internmc.facebook.com/intern/diff/D23359110)

[ghstack-poisoned]

Commit:
Update on "[FX] Support tensor-valued constants"


Tensors do not have a reliable string repr() from which they can be
constructed (and we probably don't want to rely on that, either), so
for any constant Tensor values we encounter, first search for if they
are an attribute of some module in the module hierarchy. If so, emit
a get_param to retrieve that tensor. Otherwise, we'll store away the
tensor value into a special attribute on the Module s.t. we can
retrieve it with a get_param.

Closes https://github.com/pytorch/pytorch/issues/43512

Differential Revision: [D23359110](https://our.internmc.facebook.com/intern/diff/D23359110)

[ghstack-poisoned]
